File: main/src/main/java/com/sedmelluq/discord/lavaplayer/track/playback/NonAllocatingAudioFrameBuffer.java
package com.sedmelluq.discord.lavaplayer.track.playback;

import com.sedmelluq.discord.lavaplayer.format.AudioDataFormat;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.ByteBuffer;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Audio frame buffer implementation which never allocates any new objects after creation. All instances of mutable
 * frames are preallocated, and for the data there is one byte buffer which is used as a ring buffer for the frame data.
 */
public class NonAllocatingAudioFrameBuffer extends AbstractAudioFrameBuffer {
  private static final Logger log = LoggerFactory.getLogger(NonAllocatingAudioFrameBuffer.class);

  private final int worstCaseFrameCount;
  private final ReferenceMutableAudioFrame[] frames;
  private final ReferenceMutableAudioFrame silentFrame;
  private final AtomicBoolean stopping;
  private MutableAudioFrame bridgeFrame;

  private final byte[] frameBuffer;
  private int firstFrame;
  private int frameCount;

  /**
   * @param bufferDuration The length of the internal buffer in milliseconds
   * @param format The format of the frames held in this buffer
   * @param stopping Atomic boolean which has true value when the track is in a state of pending stop.
   */
  public NonAllocatingAudioFrameBuffer(int bufferDuration, AudioDataFormat format, AtomicBoolean stopping) {
    super(format);
    int maximumFrameCount = bufferDuration / (int) format.frameDuration() + 1;
    frames = createFrames(maximumFrameCount, format);
    silentFrame = createSilentFrame(format);
    this.frameBuffer = new byte[format.expectedChunkSize() * maximumFrameCount];
    worstCaseFrameCount = frameBuffer.length / format.maximumChunkSize();
    this.stopping = stopping;
  }

  /**
   * @return Number of frames that can be added to the buffer without blocking.
   */
  @Override
  public int getRemainingCapacity() {
    synchronized (synchronizer) {
      if (frameCount == 0) {
        return worstCaseFrameCount;
      }

      int lastFrame = wrappedFrameIndex(firstFrame + frameCount - 1);

      int bufferHead = frames[firstFrame].getFrameOffset();
      int bufferTail = frames[lastFrame].getFrameEndOffset();

      int maximumFrameSize = format.maximumChunkSize();

      if (bufferHead < bufferTail) {
        return (frameBuffer.length - bufferTail) / maximumFrameSize + bufferHead / maximumFrameSize;
      } else {
        return (bufferHead - bufferTail) / maximumFrameSize;
      }
    }
  }

  /**
   * @return Total number of frames that the buffer can hold.
   */
  @Override
  public int getFullCapacity() {
    return worstCaseFrameCount;
  }

  @Override
  public void consume(AudioFrame frame) throws InterruptedException {
    // If an interrupt sent along with setting the stopping status was silently consumed elsewhere, this check should
    // still trigger. Guarantees that stopped tracks cannot get stuck in this method. Possible performance improvement:
    // offer with timeout, check stopping if timed out, then put?
    if (stopping != null && stopping.get()) {
      throw new InterruptedException();
    }

    synchronized (synchronizer) {
      if (!locked) {
        receivedFrames = true;

        if (clearOnInsert) {
          clear();
          clearOnInsert = false;
        }

        while (!attemptStore(frame)) {
          synchronizer.wait();
        }

        synchronizer.notifyAll();
      }
    }
  }

  @Override
  public AudioFrame provide() {
    synchronized (synchronizer) {
      if (provide(getBridgeFrame())) {
        return unwrapBridgeFrame();
      }

      return null;
    }
  }

  @Override
  public AudioFrame provide(long timeout, TimeUnit unit) throws TimeoutException, InterruptedException {
    synchronized (synchronizer) {
      if (provide(getBridgeFrame(), timeout, unit)) {
        return unwrapBridgeFrame();
      }

      return null;
    }
  }

  @Override
  public boolean provide(MutableAudioFrame targetFrame) {
    synchronized (synchronizer) {
      if (frameCount == 0) {
        if (terminateOnEmpty) {
          popPendingTerminator(targetFrame);
          synchronizer.notifyAll();
          return true;
        }
        return false;
      } else {
        popFrame(targetFrame);
        synchronizer.notifyAll();
        return true;
      }
    }
  }

  @Override
  public boolean provide(MutableAudioFrame targetFrame, long timeout, TimeUnit unit)
      throws TimeoutException, InterruptedException {

    long currentTime = System.nanoTime();
    long endTime = currentTime + unit.toMillis(timeout);

    synchronized (synchronizer) {
      while (frameCount == 0) {
        if (terminateOnEmpty) {
          popPendingTerminator(targetFrame);
          synchronizer.notifyAll();
          return true;
        }

        synchronizer.wait(endTime - currentTime);
        currentTime = System.nanoTime();

        if (currentTime >= endTime) {
          throw new TimeoutException();
        }
      }

      popFrame(targetFrame);
      synchronizer.notifyAll();
      return true;
    }
  }

  private void popFrame(MutableAudioFrame targetFrame) {
    ReferenceMutableAudioFrame frame = frames[firstFrame];

    if (frame.getVolume() == 0) {
      silentFrame.setTimecode(frame.getTimecode());
      frame = silentFrame;
    }

    targetFrame.setTimecode(frame.getTimecode());
    targetFrame.setVolume(frame.getVolume());
    targetFrame.setTerminator(false);
    targetFrame.store(frame.getFrameBuffer(), frame.getFrameOffset(), frame.getDataLength());

    firstFrame = wrappedFrameIndex(firstFrame + 1);
    frameCount--;
  }

  private void popPendingTerminator(MutableAudioFrame frame) {
    terminateOnEmpty = false;
    terminated = true;

    frame.setTerminator(true);
  }

  @Override
  public void clear() {
    synchronized (synchronizer) {
      frameCount = 0;
    }
  }

  @Override
  public void rebuild(AudioFrameRebuilder rebuilder) {
    log.debug("Frame rebuild not supported on non-allocating frame buffer yet.");
  }

  @Override
  public Long getLastInputTimecode() {
    synchronized (synchronizer) {
      if (!clearOnInsert && frameCount > 0) {
        return frames[wrappedFrameIndex(firstFrame + frameCount - 1)].getTimecode();
      }
    }

    return null;
  }

  private boolean attemptStore(AudioFrame frame) {
    if (frameCount >= frames.length) {
      return false;
    }

    int frameLength = frame.getDataLength();
    int frameBufferLength = frameBuffer.length;

    if (frameCount == 0) {
      firstFrame = 0;

      if (frameLength > frameBufferLength) {
        throw new IllegalArgumentException("Frame is too big for buffer.");
      }

      store(frame, 0, 0, frameLength);
    } else {
      int lastFrame = wrappedFrameIndex(firstFrame + frameCount - 1);
      int nextFrame = wrappedFrameIndex(lastFrame + 1);

      int bufferHead = frames[firstFrame].getFrameOffset();
      int bufferTail = frames[lastFrame].getFrameEndOffset();

      if (bufferHead < bufferTail) {
        if (bufferTail + frameLength <= frameBufferLength) {
          store(frame, nextFrame, bufferTail, frameLength);
        } else if (bufferHead >= frameLength) {
          store(frame, nextFrame, 0, frameLength);
        } else {
          return false;
        }
      } else if (bufferTail + frameLength <= bufferHead) {
        store(frame, nextFrame, bufferTail, frameLength);
      } else {
        return false;
      }
    }

    return true;
  }

  private int wrappedFrameIndex(int index) {
    int maximumFrameCount = frames.length;
    return index >= maximumFrameCount ? index - maximumFrameCount : index;
  }

  private void store(AudioFrame frame, int index, int frameOffset, int frameLength) {
    ReferenceMutableAudioFrame targetFrame = frames[index];
    targetFrame.setTimecode(frame.getTimecode());
    targetFrame.setVolume(frame.getVolume());
    targetFrame.setDataReference(frameBuffer, frameOffset, frameLength);

    frame.getData(frameBuffer, frameOffset);

    frameCount++;
  }

  private MutableAudioFrame getBridgeFrame() {
    if (bridgeFrame == null) {
      bridgeFrame = new MutableAudioFrame();
      bridgeFrame.setBuffer(ByteBuffer.allocate(format.maximumChunkSize()));
    }

    return bridgeFrame;
  }

  private AudioFrame unwrapBridgeFrame() {
    if (bridgeFrame.isTerminator()) {
      return TerminatorAudioFrame.INSTANCE;
    } else {
      return new ImmutableAudioFrame(bridgeFrame.getTimecode(), bridgeFrame.getData(), bridgeFrame.getVolume(),
          bridgeFrame.getFormat());
    }
  }

  private static ReferenceMutableAudioFrame[] createFrames(int frameCount, AudioDataFormat format) {
    ReferenceMutableAudioFrame[] frames = new ReferenceMutableAudioFrame[frameCount];

    for (int i = 0; i < frames.length; i++) {
      frames[i] = new ReferenceMutableAudioFrame();
      frames[i].setFormat(format);
    }

    return frames;
  }

  private static ReferenceMutableAudioFrame createSilentFrame(AudioDataFormat format) {
    ReferenceMutableAudioFrame frame = new ReferenceMutableAudioFrame();
    frame.setFormat(format);
    frame.setDataReference(format.silenceBytes(), 0, format.silenceBytes().length);
    frame.setVolume(0);
    return frame;
  }
}

File: main/src/main/java/com/sedmelluq/discord/lavaplayer/track/playback/NonAllocatingAudioFrameBuffer.kt
package com.sedmelluq.discord.lavaplayer.track.playback

import com.sedmelluq.discord.lavaplayer.format.AudioDataFormat
import com.sedmelluq.discord.lavaplayer.tools.extensions.notifyAll
import com.sedmelluq.discord.lavaplayer.tools.extensions.wait
import org.slf4j.LoggerFactory
import java.nio.ByteBuffer
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

/**
 * Audio frame buffer implementation which never allocates any new objects after creation. All instances of mutable
 * frames are pre-allocated, and for the data there is one byte buffer which is used as a ring buffer for the frame data.
 *
 * @param bufferDuration The length of the internal buffer in milliseconds
 * @param format         The format of the frames held in this buffer
 * @param stopping       Atomic boolean which has true value when the track is in a state of pending stop.
 */
class NonAllocatingAudioFrameBuffer(bufferDuration: Int, format: AudioDataFormat, private val stopping: (() -> Boolean)?) :
    AbstractAudioFrameBuffer(format) {
    /**
     * @return Total number of frames that the buffer can hold.
     */
    override val fullCapacity: Int
    private val frames: Array<ReferenceMutableAudioFrame?>
    private val silentFrame: ReferenceMutableAudioFrame
    private val frameBuffer: ByteArray
    private var bridgeFrame: MutableAudioFrame? = null
    private var firstFrame = 0
    private var frameCount = 0

    /**
     * @return Number of frames that can be added to the buffer without blocking.
     */
    override val remainingCapacity: Int
        get() {
            synchronized(synchronizer) {
                if (frameCount == 0) {
                    return fullCapacity
                }
                val lastFrame = wrappedFrameIndex(firstFrame + frameCount - 1)
                val bufferHead = frames[firstFrame]!!.frameOffset
                val bufferTail = frames[lastFrame]!!.frameEndOffset
                val maximumFrameSize = format.maximumChunkSize
                return if (bufferHead < bufferTail) {
                    (frameBuffer.size - bufferTail) / maximumFrameSize + bufferHead / maximumFrameSize
                } else {
                    (bufferHead - bufferTail) / maximumFrameSize
                }
            }
        }

    init {
        val maximumFrameCount = bufferDuration / format.frameDuration().toInt() + 1
        frames = createFrames(maximumFrameCount, format)
        silentFrame = createSilentFrame(format)
        frameBuffer = ByteArray(format.expectedChunkSize * maximumFrameCount)
        fullCapacity = frameBuffer.size / format.maximumChunkSize
    }

    @Throws(InterruptedException::class)
    override fun consume(frame: AudioFrame) {
        // If an interrupt sent along with setting the stopping status was silently consumed elsewhere, this check should
        // still trigger. Guarantees that stopped tracks cannot get stuck in this method. Possible performance improvement:
        // offer with timeout, check stopping if timed out, then put?
        if (stopping?.invoke() == true) {
            throw InterruptedException()
        }

        synchronized(synchronizer) {
            if (!locked) {
                receivedFrames = true
                if (clearOnInsert) {
                    clear()
                    clearOnInsert = false
                }
                while (!attemptStore(frame)) {
                    synchronizer.wait()
                }
                synchronizer.notifyAll()
            }
        }
    }

    override fun provide(): AudioFrame? {
        synchronized(synchronizer) {
            return if (provide(getBridgeFrame())) {
                unwrapBridgeFrame()
            } else null
        }
    }

    @Throws(TimeoutException::class, InterruptedException::class)
    override fun provide(timeout: Long, unit: TimeUnit): AudioFrame? {
        synchronized(synchronizer) {
            return if (provide(getBridgeFrame(), timeout, unit)) {
                unwrapBridgeFrame()
            } else null
        }
    }

    override fun provide(targetFrame: MutableAudioFrame): Boolean {
        synchronized(synchronizer) {
            return if (frameCount == 0) {
                if (terminateOnEmpty) {
                    popPendingTerminator(targetFrame)
                    synchronizer.notifyAll()
                    return true
                }
                false
            } else {
                popFrame(targetFrame)
                synchronizer.notifyAll()
                true
            }
        }
    }

    @Throws(TimeoutException::class, InterruptedException::class)
    override fun provide(targetFrame: MutableAudioFrame, timeout: Long, unit: TimeUnit): Boolean {
        var currentTime = System.nanoTime()
        val endTime = currentTime + unit.toMillis(timeout)
        synchronized(synchronizer) {
            while (frameCount == 0) {
                if (terminateOnEmpty) {
                    popPendingTerminator(targetFrame)
                    synchronizer.notifyAll()
                    return true
                }
                synchronizer.wait(endTime - currentTime)
                currentTime = System.nanoTime()
                if (currentTime >= endTime) {
                    throw TimeoutException()
                }
            }
            popFrame(targetFrame)
            synchronizer.notifyAll()
            return true
        }
    }

    private fun popFrame(targetFrame: MutableAudioFrame) {
        var frame = frames[firstFrame]
        if (frame!!.volume == 0) {
            silentFrame.timecode = frame.timecode
            frame = silentFrame
        }
        targetFrame.timecode = frame.timecode
        targetFrame.volume = frame.volume
        targetFrame.isTerminator = false
        targetFrame.store(frame.frameBuffer, frame.frameOffset, frame.dataLength)
        firstFrame = wrappedFrameIndex(firstFrame + 1)
        frameCount--
    }

    private fun popPendingTerminator(frame: MutableAudioFrame) {
        terminateOnEmpty = false
        terminated = true
        frame.isTerminator = true
    }

    override fun clear() {
        synchronized(synchronizer) { frameCount = 0 }
    }

    override fun rebuild(rebuilder: AudioFrameRebuilder) {
        log.debug("Frame rebuild not supported on non-allocating frame buffer yet.")
    }

    override val lastInputTimecode: Long?
        get() {
            synchronized(synchronizer) {
                if (!clearOnInsert && frameCount > 0) {
                    return frames[wrappedFrameIndex(firstFrame + frameCount - 1)]!!.timecode
                }
            }
            return null
        }

    private fun attemptStore(frame: AudioFrame): Boolean {
        if (frameCount >= frames.size) {
            return false
        }
        val frameLength = frame.dataLength
        val frameBufferLength = frameBuffer.size
        if (frameCount == 0) {
            firstFrame = 0
            require(frameLength <= frameBufferLength) { "Frame is too big for buffer." }
            store(frame, 0, 0, frameLength)
        } else {
            val lastFrame = wrappedFrameIndex(firstFrame + frameCount - 1)
            val nextFrame = wrappedFrameIndex(lastFrame + 1)
            val bufferHead = frames[firstFrame]!!.frameOffset
            val bufferTail = frames[lastFrame]!!.frameEndOffset
            if (bufferHead < bufferTail) {
                if (bufferTail + frameLength <= frameBufferLength) {
                    store(frame, nextFrame, bufferTail, frameLength)
                } else if (bufferHead >= frameLength) {
                    store(frame, nextFrame, 0, frameLength)
                } else {
                    return false
                }
            } else if (bufferTail + frameLength <= bufferHead) {
                store(frame, nextFrame, bufferTail, frameLength)
            } else {
                return false
            }
        }
        return true
    }

    private fun wrappedFrameIndex(index: Int): Int {
        val maximumFrameCount = frames.size
        return if (index >= maximumFrameCount) index - maximumFrameCount else index
    }

    private fun store(frame: AudioFrame, index: Int, frameOffset: Int, frameLength: Int) {
        val targetFrame = frames[index]
        targetFrame!!.timecode = frame.timecode
        targetFrame.volume = frame.volume
        targetFrame.setDataReference(frameBuffer, frameOffset, frameLength)
        frame.getData(frameBuffer, frameOffset)
        frameCount++
    }

    private fun getBridgeFrame(): MutableAudioFrame {
        if (bridgeFrame == null) {
            bridgeFrame = MutableAudioFrame()
            bridgeFrame!!.setBuffer(ByteBuffer.allocate(format.maximumChunkSize))
        }
        return bridgeFrame!!
    }

    private fun unwrapBridgeFrame(): AudioFrame {
        return if (bridgeFrame!!.isTerminator) {
            TerminatorAudioFrame
        } else {
            ImmutableAudioFrame(
                bridgeFrame!!.timecode,
                bridgeFrame!!.data,
                bridgeFrame!!.volume,
                bridgeFrame!!.format!!
            )
        }
    }

    override fun signalWaiters() {
        synchronized(synchronizer) { synchronizer.notifyAll() }
    }

    companion object {
        private val log = LoggerFactory.getLogger(NonAllocatingAudioFrameBuffer::class.java)
        private fun createFrames(frameCount: Int, format: AudioDataFormat): Array<ReferenceMutableAudioFrame?> {
            val frames = arrayOfNulls<ReferenceMutableAudioFrame>(frameCount)
            for (i in frames.indices) {
                frames[i] = ReferenceMutableAudioFrame()
                frames[i]!!.format = format
            }
            return frames
        }

        private fun createSilentFrame(format: AudioDataFormat): ReferenceMutableAudioFrame {
            val frame = ReferenceMutableAudioFrame()
            frame.format = format
            frame.setDataReference(format.silenceBytes(), 0, format.silenceBytes().size)
            frame.volume = 0
            return frame
        }
    }
}
